File: pycode/problem_solving/queen_attack/queen_attack.py
#!/bin/python3

import math
import os
import random
import re
import sys

#
# Complete the 'queensAttack' function below.
#
# The function is expected to return an INTEGER.
# The function accepts following parameters:
#  1. INTEGER n
#  2. INTEGER k
#  3. INTEGER r_q
#  4. INTEGER c_q
#  5. 2D_INTEGER_ARRAY obstacles

def queensAttack(n, k, r_q, c_q, obstacles):
    # Write your code here
    # generae eight rays from queen position
    obstacles = [str(x) for x in obstacles]
    obstacles = set(obstacles)
    count = 0
    # down
    for offset in range(1, r_q): 
        cur = str([r_q - offset, c_q] )
        if (cur in obstacles):
            break
        count += 1

    # up
    for offset in range(1, n-r_q+1): 
        cur = str([r_q + offset, c_q] )
        if (cur in obstacles):
            break
        count += 1

    # left
    for offset in range(1, c_q): 
        cur = str([r_q, c_q - offset] )
        if (cur in obstacles):
            break
        count += 1

    # right 
    for offset in range(1, n-c_q+1): 
        cur = str([r_q, c_q + offset] )
        if (cur in obstacles):
            break
        count += 1

    # ne 
    mq = max(r_q, c_q)
    for offset in range(1, n-mq+1): 
        cur = str([r_q + offset, c_q + offset] )
        if (cur in obstacles):
            break
        print(cur)
        count += 1

    # sw 
    mq = min(r_q, c_q)
    for offset in range(1, mq): 
        cur = str([r_q - offset, c_q - offset] )
        if (cur in obstacles):
            break
        print(cur)
        count += 1

    # se 
    mq = min(r_q, n-c_q+1)
    for offset in range(1, mq): 
        cur = str([r_q - offset, c_q + offset] )
        if (cur in obstacles):
            break
        print(cur)
        count += 1

    # nw 
    mq = min(n-r_q+1, c_q)
    for offset in range(1, mq): 
        cur = str([r_q + offset, c_q - offset] )
        if (cur in obstacles):
            break
        print(cur)
        count += 1

    return count



if __name__ == '__main__':
    with open('./data/input.txt', 'r') as sys.stdin:

        first_multiple_input = input().rstrip().split()
        # board size
        n = int(first_multiple_input[0])
        # num obstacles
        k = int(first_multiple_input[1])

        second_multiple_input = input().rstrip().split()

        # queen position
        r_q = int(second_multiple_input[0])
        c_q = int(second_multiple_input[1])

        obstacles = []

        # read obstables
        for _ in range(k):
            obstacles.append(list(map(int, input().rstrip().split())))

        # print(obstacles)

        result = queensAttack(n, k, r_q, c_q, obstacles)
        print(result)

File: jscode/queen_attack/queen_attack.js
'use strict';

const fs = require('fs');

process.stdin.resume();
process.stdin.setEncoding('utf-8');

let inputString = '';
let currentLine = 0;

process.stdin.on('data', function(inputStdin) {
    inputString += inputStdin;
});

process.stdin.on('end', function() {
    inputString = inputString.split('\n');

    main();
});

function readLine() {
    return inputString[currentLine++];
}

/*
 * Complete the 'queensAttack' function below.
 *
 * The function is expected to return an INTEGER.
 * The function accepts following parameters:
 *  1. INTEGER n
 *  2. INTEGER k
 *  3. INTEGER r_q
 *  4. INTEGER c_q
 *  5. 2D_INTEGER_ARRAY obstacles
 */

const DEBUG = 0;
function show(x) {
    if (DEBUG == 1) console.log(x)
}

function queensAttack(n, k, r_q, c_q, obstacles) {
    // Write your code here
    obstacles = new Set(obstacles.map(x => x.toString()))
    let count = 0;
    // down
    for (var k=r_q-1; k>0; k--) {
        var cur = [k, c_q].toString()
        if (obstacles.has(cur)) {
            break
        } else {
            count++;
        }
    }

    // up 
    for (var k=r_q+1; k<=n; k++) {
        var cur = [k, c_q].toString()
        if (obstacles.has(cur)) {
            break
        } else {
            count++;
        }
    }

    // left 
    for (var k=c_q-1; k>0; k--) {
        var cur = [r_q, k].toString()
        if (obstacles.has(cur)) {
            break
        } else {
            count++;
        }
    }

    // right
    for (var k=c_q+1; k<=n; k++) {
        var cur = [r_q, k].toString()
        if (obstacles.has(cur)) {
            break
        } else {
            count++;
        }
    }

    // NE
    for (var k=r_q+1, j=c_q+1; k<=n && j<=n; k++, j++) {
        var cur = [k, j].toString()
        if (obstacles.has(cur)) {
            break
        } else {
            count++;
        }
    }

    // SW
    for (var k=r_q-1, j=c_q-1; k>0 && j>0; k--, j--) {
        var cur = [k, j].toString()
        if (obstacles.has(cur)) {
            break
        } else {
            count++;
        }
    }

    // SE
    for (var k=r_q-1, j=c_q+1; k>0 && j<=n; k--, j++) {
        var cur = [k, j].toString()
        if (obstacles.has(cur)) {
            break
        } else {
            count++;
        }
    }

    // NW 
    for (var k=r_q+1, j=c_q-1; k<=n && j>0; k++, j--) {
        var cur = [k, j].toString()
        if (obstacles.has(cur)) {
            break
        } else {
            count++;
        }
    }
    return count

}

function main() {
    // const ws = fs.createWriteStream(process.env.OUTPUT_PATH);
    // const ws = fs.createWriteStream("./data/input.txt")

    const firstMultipleInput = readLine().replace(/\s+$/g, '').split(' ');
    const n = parseInt(firstMultipleInput[0], 10);
    const k = parseInt(firstMultipleInput[1], 10);

    const secondMultipleInput = readLine().replace(/\s+$/g, '').split(' ');
    const r_q = parseInt(secondMultipleInput[0], 10);
    const c_q = parseInt(secondMultipleInput[1], 10);

    let obstacles = Array(k);

    for (let i = 0; i < k; i++) {
        obstacles[i] = readLine().replace(/\s+$/g, '').split(' ').map(obstaclesTemp => parseInt(obstaclesTemp, 10));
    }

    const result = queensAttack(n, k, r_q, c_q, obstacles);
    console.log(result)

    // ws.write(result + '\n');
    // ws.end();
}
